textfsm.py

```diff
@@ -171,13 +171,37 @@ class Key(OptionBase):
     """Value constitutes part of the Key of the record."""
 
   class List(OptionBase):
-    """Value takes the form of a list."""
+    """
+    Value takes the form of a list.
+
+    If the value regex contains mested match groups in the form (?P<name>regex),
+    instead of adding a string to the list, we add a dictionary of the groups.
+
+    Eg.
+    Value List ((?P<name>\w+)\s+(?P<age>\d+)) would create results like [{'name': 'Bob', 'age': 32}]
+
+    Do not give nested groups the same name as another values in the template.
+    """
 
     def OnCreateOptions(self):
       self.OnClearAllVar()
 
     def OnAssignVar(self):
-      self._value.append(self.value.value)
+      # Store the compiled regex object after compiling it only once
+      if not getattr(self.value, 'compiled_regex', None):
```

Comment:
I'd be inclined to compile and insert the regexp in the Value object directly. An object should only have attributes added from within the class (this can be added where 'template' is assigned).

As a bonus, the use of duplicate group names will be caught at template parse time rather than if/when the match takes place.